src/dispatch/batch.rs

```diff
@@ -0,0 +1,453 @@
+/// write resources of the `System`s contained by the `Batch` ("sub systems").
+#[derive(Debug)]
+pub struct BatchAccessor {
+    reads: Vec<ResourceId>,
+    writes: Vec<ResourceId>,
+}
+
+impl BatchAccessor {
+    /// Create the `BatchAccessor`
+    pub fn new(reads: Vec<ResourceId>, writes: Vec<ResourceId>) -> Self {
+        BatchAccessor { reads, writes }
+    }
+}
+
+impl Accessor for BatchAccessor {
+    fn try_new() -> Option<Self> {
+        None
+    }
+
+    fn reads(&self) -> Vec<ResourceId> {
+        self.reads.clone()
+    }
+
+    fn writes(&self) -> Vec<ResourceId> {
+        self.writes.clone()
+    }
+}
+
+/// The `BatchUncheckedWorld` wrap an instance of the world.
+/// You have to specify this as `SystemData` of the `BatchControllerSystem`.
+pub struct BatchUncheckedWorld<'a>(pub &'a World);
+
+impl<'a> DynamicSystemData<'a> for BatchUncheckedWorld<'a> {
+    type Accessor = BatchAccessor;
+
+    fn setup(_accessor: &Self::Accessor, _world: &mut World) {}
+
+    fn fetch(_access: &Self::Accessor, world: &'a World) -> Self {
+        BatchUncheckedWorld(world)
+    }
+}
+
+/// The `BatchController` is the additional trait that a normal System must
+/// implement in order to be used as `BatchControllerSystem`.
+///
+/// Note that the `System` must also implement `Send` because the `Dispatcher`
+/// is by default un-send.
+/// Is safe to implement `Send` and `Sync` because the `BatchAccessor` keep
+/// tracks of all used resources and thus the `System` can be safely executed in
+/// multi thread.
+pub trait BatchController<'a, 'b> {
+    /// Here you must set all the `Resource` types that you want to use directly
+    /// inside the `BatchControllerSystem`.
+    ///
+    /// These have to be specified here, instead of `SystemData` (as
+    /// a normal `System` does) because there is a direct dependecy with the sub
+    /// `System`s `Resource`.
+    ///
+    /// Since the sub `System`s will use the same `Resource' of the
+    /// `BatchControllerSystem` is necessary to drop the references to the
+    /// fetched `Resource` before to dispatch the inner `Dispatcher`.
+    ///
+    /// Now is easy to understand that specify the `BatchControllerSystem`
+    /// `Resource` in the `SystemData` doesn't allow to drop the reference
+    /// before the sub dispatching resulting in `Panic`.
+    ///
+    /// So this mechanism allow you to specify the `Resource` that you will use
+    /// in the `BatchControllerSystem` then is safe to fetch them directly
+    /// from the world (As the example  "examples/batch_dispatching.rs" show).
+    ///
+    /// Note that it's not required to specify the sub systems resources here
+    /// because they are handled automatically.
+    type BatchSystemData: SystemData<'a>;
+
+    /// Create the instance of the `BatchControllerSystem`
+    ///
+    /// This function is unsafe because it depends from the `BatchAccessor` that
+    /// if created wrongly can Panics.
```

Comment:
Panicking is not memory unsafe in general and so I am still not sure what is going on with this whole `Send` business. 
@torkleyy Is there reason for not implementing `Send` for `Dispatcher`?